rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,677 @@
+---
+feature: nix-language-version
+start-date: 2022-12-12
+author: @fricklerhandwerk @yorickvp
+co-authors: @thufschmitt @Ericson2314 @infinisil
+shepherd-team:
+shepherd-leader:
+related-issues: https://github.com/NixOS/nix/issues/7255
+---
+
+# RFC 137 – Nix language versioning
+
+# Summary
+[summary]: #summary
+
+Introduce a convention to determine which version of the Nix language grammar to use for parsing and evaluating Nix expressions.
+
+Add parameters to the Nix language evaluator, controlling the behavior of deprecation warnings and errors.
+
+# Motivation
+[motivation]: #motivation
+
+The stability of Nix language has been praised on multiple occasions, e.g. [Nix and legacy enterprise software development: an unlikely match made in heaven](https://talks.nixcon.org/nixcon-2022/talk/QQPBFW/).
+Yet, as with any software system, in order to accommodate new insights, we want to allow the Nix language to evolve.
+This sometimes involves backward-incompatible ("breaking") changes that currently cannot be made without significant downstream disruption.
+
+Therefore we propose a mechanims and policies to introduce changes to the Nix language in a controlled and deliberate manner.
+It aims to avoid breaking existing code, to prevent inadvertently breaking reproducibility, and to minimise maintenance burden for implementors and users.
+
+## Motivating examples
+
+Incompatible changes from the past:
+
+- [A changelog of Nix (language) versions](https://code.tvl.fyi/about/tvix/docs/lang-version.md), as reflected in `builtins.langVersion`
+- There have been other, sometimes breaking changes to the language that have not resulted in an increment of the language version (e.g. the recent `fetchGit` changes).
+- The `builtins.toJSON 1.000001` output [changed in Nix 2.12](https://github.com/NixOS/nix/issues/8259).
+
+Possbile future changes that are in discussion:
```

Comment:
```suggestion
Possible future changes that are in discussion:
```

Reply:
I'd suggest not doing typo suggestions in RFC threads, since these kind of things are really not important to the end result, so it's mostly noise compared to reviewing more important things.

Reply:
Alright, it was a very easy one while I was reading :).

Reply:
I think it's fine. Readers can assume that closed comments are incorporated in the RFC text.

Reply:
I agree. Fixing typos can also make it easier for the next reader to understand. Especially in cases where the intended word isn't obvious. And I don't want to be making judgements on what typos are worth fixing. I think it is fine to highlight them all as ideally they would all be caught and corrected before the RFC is merged.